crates/emitter/src/ast_emitter.rs

```diff
@@ -481,102 +482,94 @@ impl<'alloc, 'opt> AstEmitter<'alloc, 'opt> {
+                    }
+                    .emit(self)?;
                 }
             },
             _ => return Err(EmitError::NotImplemented("TODO: non data property")),
         }
-
         Ok(())
     }
 
-    fn to_property_index(&self, index: SourceAtomSetIndex) -> Option<u32> {
-        let s = self.compilation_info.atoms.get(index);
-        s.parse::<u32>().ok()
-    }
-
     fn emit_array_expression(&mut self, array: &ArrayExpression) -> Result<(), EmitError> {
-        // Initialize the array to its minimum possible length.
-        let min_length = array
-            .elements
-            .iter()
-            .map(|e| match e {
-                ArrayExpressionElement::Expression(_) => 1,
-                ArrayExpressionElement::Elision { .. } => 1,
-                ArrayExpressionElement::SpreadElement { .. } => 0,
-            })
-            .sum::<u32>();
-
-        self.emit.new_array(min_length);
-
-        for (index, element) in array.elements.iter().enumerate() {
-            match element {
-                ArrayExpressionElement::Expression(expr) => {
-                    self.emit_expression(&expr)?;
-                    self.emit.init_elem_array(index as u32);
-                }
-                ArrayExpressionElement::Elision { .. } => {
-                    self.emit.hole();
-                    self.emit.init_elem_array(index as u32);
-                }
-                ArrayExpressionElement::SpreadElement { .. } => {
-                    return Err(EmitError::NotImplemented("TODO: SpreadElement"));
+        ArrayEmitter {
+            elements: array.elements.iter(),
+            elem_kind: |e| match e {
+                ArrayExpressionElement::Expression(..) => ArrayElementKind::Normal,
+                ArrayExpressionElement::Elision { .. } => ArrayElementKind::Elision,
+                ArrayExpressionElement::SpreadElement(..) => ArrayElementKind::Spread,
+            },
+            elem: |emitter, state, elem| {
+                match elem {
+                    ArrayExpressionElement::Expression(expr) => {
+                        ArrayElementEmitter {
+                            state,
+                            elem: |emitter| emitter.emit_expression(expr),
+                        }
+                        .emit(emitter)?;
+                    }
+                    ArrayExpressionElement::Elision { .. } => {
+                        ArrayElisionEmitter { state }.emit(emitter);
+                    }
+                    ArrayExpressionElement::SpreadElement(expr) => {
+                        ArraySpreadEmitter {
+                            state,
+                            elem: |emitter| emitter.emit_expression(expr),
+                        }
+                        .emit(emitter)?;
+                    }
                 }
-            }
+                Ok(())
+            },
         }
+        .emit(self)?;
 
         Ok(())
```

Comment:
Not sure what the best practice is, but do we still need this ok? it looks like ArrayEmitter::emit returns `ok` as well.